### PR TITLE
[TACACS+] Add audisp-tacplus for per-command accounting.

### PR DESCRIPTION
**Pull request by liuh-80**
This pull request integrate audisp-tacplus to SONiC for per-command accounting.

#### Why I did it
To support TACACS per-command accounting, we integrate audisp-tacplus project to sonic.

#### How I did it
1. Add auditd service to SONiC
2. Port and patch audisp-tacplus to SONiC

#### How to verify it
UT with CUnit to cover all new code in usersecret-filter.c
Also pass all current UT.

#### Which release branch to backport (provide reason below if selected)
N/A

#### Description for the changelog
Add audisp-tacplus for per-command accounting.

### CODE REVIEW

**Comment by liuh-80** [comment]
This just draft PR still ongoing.
Need new code to remov user secret from accounting message.
Also need local accounting with syslog support.

**Comment by liuh-80** [comment]
Not ready for review, publish by mistake, switch back to draft. still need following change:
1. Remove user secret from accounting message.
2. Support syslog as 'local' accounting
3. Make accounting method 'local' & 'tacacs' configurable.

File: src/tacacs/audisp/Makefile [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+	# Obtain audisp-tacplus
+	rm -rf ./audisp-tacplus
+
+	git clone https://github.com/daveolson53/audisp-tacplus.git
+
+	# checkout by sha1
+	pushd ./pam_tacplus
+	git checkout 559c9f22edd4f2dea0ecedffb3ad9502b12a75b6
```
Comment (qiluo-msft) [inline src/tacacs/audisp/Makefile:17]: >559c9f22edd4f2dea0ecedffb3ad9502b12a75b6 [](http://example.com/codeflow?start=14&length=40)

How do you choose this commit? Is it on specific branch?

File: src/tacacs/audisp/patches/0001-Port-audisp-tacplus-to-sonic.patch [hunk 2]
```diff
@@ -0,0 +1,423 @@
+From b210fb084270aaf95ab10a917fd06c367906e85c Mon Sep 17 00:00:00 2001
+From: liuh-80 <58683130+liuh-80@users.noreply.github.com>
+Date: Tue, 14 Sep 2021 15:53:25 +0800
+Subject: [PATCH] Port audisp-tacplus to sonic.
```
Comment (qiluo-msft) [inline src/tacacs/audisp/patches/0001-Port-audisp-tacplus-to-sonic.patch:4]: >sonic [](http://example.com/codeflow?start=40&length=5)

I don't see anything related to sonic. Could you give concrete title? Does it really need to know SONiC?

File: src/tacacs/audisp/Makefile (hunk 1 above)
Comment (liuh-80) [reply src/tacacs/audisp/Makefile:17]: This commit is the latest commit on master branch.
Our patch will base on this commit.

File: src/tacacs/audisp/patches/series [hunk 3]
```diff
@@ -0,0 +1,3 @@
+0001-Port-audisp-tacplus-to-sonic.patch
```
Comment (qiluo-msft) [inline src/tacacs/audisp/patches/series:1]: Do you need to change patch name in series file?

File: rules/tacacs.dep [hunk 4]
```diff
@@ -21,3 +21,15 @@ $(LIBNSS_TACPLUS)_CACHE_MODE  := GIT_CONTENT_SHA
 $(LIBNSS_TACPLUS)_DEP_FLAGS   := $(SONIC_COMMON_FLAGS_LIST)
 $(LIBNSS_TACPLUS)_DEP_FILES   := $(DEP_FILES)
 
+
+
+
+SPATH       := $($(BASH_TACPLUS)_SRC_PATH)
```
Comment (qiluo-msft) [inline rules/tacacs.dep:27]: too many empty lines

File: src/tacacs/audisp/Makefile [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	# checkout by sha1
+	pushd ./audisp-tacplus
+	git checkout 559c9f22edd4f2dea0ecedffb3ad9502b12a75b6
+
+	# Apply patches
+	cp -r ../patches patches
+	quilt push -a
+
+	# fix aclocal depency issue by run autoconf
+	autoreconf
```
Comment (qiluo-msft) [inline src/tacacs/audisp/Makefile:22]: >autoreconf [](http://example.com/codeflow?start=1&length=10)

Is it same as [https://github.com/daveolson53/audisp-tacplus/blob/master/auto.sh](https://github.com/daveolson53/audisp-tacplus/blob/master/auto.sh)?

Comment (liuh-80) [reply src/tacacs/audisp/Makefile:22]: Yes, the original 'configure' file depends on a older version of aclocal, so we need run autoconf first to fix the aclocal version issue.

Comment (qiluo-msft) [reply src/tacacs/audisp/Makefile:22]: Then use `./auto.sh` ?

File: rules/tacacs.dep (hunk 4 above)
Comment (qiluo-msft) [reply rules/tacacs.dep:27]: Still applicable

**Review by qiluo-msft — APPROVED** [review]

**Comment by liuh-80** [comment]
Create this PR for better code review: https://github.com/liuh-80/audisp-tacplus/pull/1

File: src/tacacs/audisp/Makefile (hunk 5 above)
Comment (liuh-80) [reply src/tacacs/audisp/Makefile]: Fixed.

**Comment by qiluo-msft** [comment]
I close all my comment and focus on the patches.

**Comment by qiluo-msft** [comment]
Please resolve the conflict.

**Comment by liuh-80** [comment]
> Please resolve the conflict.
Conflict fixed.

**Review by qiluo-msft — DISMISSED** [review]

**Review by xumia — APPROVED** [review]

**Comment by liuh-80** [comment]
This PR build break because file permission of  build_debian.sh change to 644, so fix this by:
'git update-index --chmod=+x build_debian.sh' and submit a new commit.

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**